python/src/internal.rs

```diff
@@ -239,6 +251,48 @@ impl From<&Instant> for HudiInstant {
     }
 }
 
+#[cfg(not(tarpaulin_include))]
+#[pyclass(name = "HudiDataSource")]
+#[derive(Clone)]
+pub struct HudiDataSource {
+    table: InternalHudiDataSource,
```

Comment:
in rust, we have hudi-datafusion crate, so HudiDataSource is meant for datafusion. 
In python, we need to make the name explicit for datafusion so people don't confuse it with datasource for other integrations (in future). 
Also can we make all related functionalities in this PR available only when the python `datafusion` feature is enabled? like people use `pip install hudi[datafusion]` to access the related features.

Reply:
I renamed HudiDataSource to HudiDataFusionDataSource, also make datafusion dependency optional in the pyproject.toml so people can use `pip install hudi[datafusion]`.

Additionally I extracted HudiDataFusionDataSource(datafusion_internal.rs) to a different module so we can enable or disable it with rust feature flags.